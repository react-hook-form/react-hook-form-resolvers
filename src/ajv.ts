import { appendErrors } from 'react-hook-form';
import { ValidationError, ValidateFunction } from 'ajv';

type FieldValues = Record<string, any>;

const parseErrorSchema = (
  validationError: ValidationError,
  validateAllFieldCriteria: boolean,
) =>
  Array.isArray(validationError.errors)
    ? validationError.errors.reduce(
        (
          previous: FieldValues,
          { dataPath, message = '', params, propertyName = '' },
        ) => {
          const path =
            dataPath.replace(/\//g, '.').replace(/^\./, '') || propertyName;
          return {
            ...previous,
            ...(path
              ? previous[path] && validateAllFieldCriteria
                ? {
                    [path]: appendErrors(
                      path,
                      validateAllFieldCriteria,
                      previous,
                      '',
                      message,
                    ),
                  }
                : {
                    [path]: previous[path] || {
                      message,
                      params,
                      ...(validateAllFieldCriteria
                        ? {
                            types: {
                              ['']: message || true,
                            },
                          }
                        : {}),
                    },
                  }
              : {}),
          };
        },
        {},
      )
    : [];

interface schema {
  validate: ValidateFunction;
}

export const ajvResolver = (validationSchema: schema | ValidateFunction) => {
  const validate =
    typeof (validationSchema as schema)?.validate === 'function'
      ? (validationSchema as schema).validate
      : typeof validationSchema === 'function'
      ? validationSchema
      : undefined;

  if (!validate) {
    throw new Error('Invalid AJV validate function');
  }

  return async (data: any, _: any = {}, validateAllFieldCriteria = false) => {
    try {
      await validate(data);
      return { values: data, errors: {} };
    } catch (e) {
      return {
        values: {},
        errors: parseErrorSchema(e, validateAllFieldCriteria),
      };
    }
  };
};
