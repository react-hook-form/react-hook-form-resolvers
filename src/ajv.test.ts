import * as Ajv from 'ajv';
import { ajvResolver } from './ajv';

const schema = {
  $async: true,
  type: 'object',
  properties: {
    name: {
      type: 'string',
      pattern: '[a-zA-Z]',
      minLength: 3,
    },
    age: {
      type: 'integer',
      minimum: 0,
    },
    email: {
      type: 'string',
      format: 'email',
    },
    hobbies: {
      type: 'array',
      minItems: 1,
      items: {
        properties: {
          description: {
            type: 'string',
            minLength: 3,
          },
        },
        required: ['description'],
      },
    },
  },
  required: ['name', 'age', 'email', 'hobbies'],
};

describe('ajvResolver', () => {
  describe('using Ajv.compile', () => {
    let validate: any;

    beforeEach(() => {
      validate = new Ajv({ allErrors: true, coerceTypes: true }).compile(
        schema,
      );
    });

    it('should get values', async () => {
      const data = {
        name: 'jimmy',
        age: '24',
        email: 'jimmy@jimmy.com',
        hobbies: ['tennis'],
      };

      expect(await ajvResolver(validate)(data)).toEqual({
        errors: {},
        values: data,
      });
    });

    it('should get errors for array fields', async () => {
      const data = {
        name: 'jimmy',
        age: 24,
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: '',
          },
        ],
      };
      expect(await ajvResolver(validate)(data)).toMatchSnapshot();
    });

    it('should get errors for top level properties', async () => {
      const data = {
        name: 24,
        age: 'jimmy',
        email: 'jimmyjimmy.com',
        hobbies: [
          {
            description: 'Tennis',
          },
        ],
      };

      expect(await ajvResolver(validate)(data)).toMatchSnapshot();
    });

    it('should return empty object when validate pass', async () => {
      expect(
        await ajvResolver({
          validate: () => new Promise((resolve) => resolve()) as any,
        })({}),
      ).toEqual({ errors: {}, values: {} });
    });
  });

  describe('using Ajv.validate', () => {
    let validate: any;

    beforeEach(() => {
      const ajv = new Ajv({ allErrors: true, coerceTypes: true });
      validate = (data: any) => ajv.validate(schema, data);
    });

    it('should get values', async () => {
      const data = {
        name: 'jimmy',
        age: '24',
        email: 'jimmy@jimmy.com',
        hobbies: ['tennis'],
      };

      expect(await ajvResolver(validate)(data)).toEqual({
        errors: {},
        values: data,
      });
    });

    it('should get errors for array fields', async () => {
      const data = {
        name: 'jimmy',
        age: 24,
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: '',
          },
        ],
      };
      expect(await ajvResolver(validate)(data)).toMatchSnapshot();
    });

    it('should get errors for top level properties', async () => {
      const data = {
        name: 24,
        age: 'jimmy',
        email: 'jimmyjimmy.com',
        hobbies: [
          {
            description: 'Tennis',
          },
        ],
      };

      expect(await ajvResolver(validate)(data)).toMatchSnapshot();
    });

    it('should return empty object when validate pass', async () => {
      expect(
        await ajvResolver({
          validate: () => new Promise((resolve) => resolve()) as any,
        })({}),
      ).toEqual({ errors: {}, values: {} });
    });
  });
});
