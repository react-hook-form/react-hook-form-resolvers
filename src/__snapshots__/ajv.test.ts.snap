// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ajvResolver using Ajv.compile should get errors for array fields 1`] = `
Object {
  "errors": Object {
    "hobbies[0].description": Object {
      "message": "should NOT be shorter than 3 characters",
      "params": Object {
        "limit": 3,
      },
    },
  },
  "values": Object {},
}
`;

exports[`ajvResolver using Ajv.compile should get errors for top level properties 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "should be integer",
      "params": Object {
        "type": "integer",
      },
    },
    "email": Object {
      "message": "should match format \\"email\\"",
      "params": Object {
        "format": "email",
      },
    },
    "name": Object {
      "message": "should NOT be shorter than 3 characters",
      "params": Object {
        "limit": 3,
      },
    },
  },
  "values": Object {},
}
`;

exports[`ajvResolver using Ajv.validate should get errors for array fields 1`] = `
Object {
  "errors": Object {
    "hobbies[0].description": Object {
      "message": "should NOT be shorter than 3 characters",
      "params": Object {
        "limit": 3,
      },
    },
  },
  "values": Object {},
}
`;

exports[`ajvResolver using Ajv.validate should get errors for top level properties 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "should be integer",
      "params": Object {
        "type": "integer",
      },
    },
    "email": Object {
      "message": "should match format \\"email\\"",
      "params": Object {
        "format": "email",
      },
    },
    "name": Object {
      "message": "should NOT be shorter than 3 characters",
      "params": Object {
        "limit": 3,
      },
    },
  },
  "values": Object {},
}
`;
